src/python/pants/core/util_rules/pants_environment.py

```diff
@@ -0,0 +1,65 @@
+# Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+import logging
+import os
+import re
+from dataclasses import dataclass
+from typing import Dict, Optional, Sequence, cast
+
+from pants.engine.rules import side_effecting
+from pants.util.frozendict import FrozenDict
+from pants.util.meta import frozen_after_init
+
+logger = logging.getLogger(__name__)
+
+
+@side_effecting
+@frozen_after_init
+@dataclass(unsafe_hash=True)
+class PantsEnvironment:
+    """PantsEnvironment is a representation of the environment variables the currently-executing
+    pants process was invoked with.
+
+    Since the exact contents of the environment are very specific to a given invocation of pants,
+    this type is marked as side-effecting, and can only be requested as an input to an uncacheable
+    rule. An uncacheable rule should accept this type as an input, and use `get_subset` to filter
+    the environment rules down to only the ones other rules care about as inputs, in order to avoid
+    a lot of cache invalidation if the environment changes spuriously.
+    """
+
+    env: FrozenDict[str, str]
+
+    def __init__(self, env: Optional[Dict[str, str]] = None):
+        """Initialize a `PantsEnvironment` with the current contents of the environment."""
+        if env:
+            self.env = FrozenDict(env)
+        else:
+            self.env = FrozenDict(dict(os.environ))
+
+    @classmethod
+    def create_mock(cls, mock_env: Dict[str, str]) -> "PantsEnvironment":
+        """This method is intended to be used in tests only to create a mock `PantsEnvironment`."""
+        frozen = FrozenDict(mock_env)
+        return cls(env=cast(Optional[Dict[str, str]], frozen))
+
+    def get_subset(self, requested: Sequence[str]) -> FrozenDict[str, Optional[str]]:
+        """Given a list of extra environment variable specifiers as strings, filter the contents of
+        the pants environment to only the variables that were specified as a pants configuration
+        option."""
+
+        name_value_re = re.compile(r"(\w+)=(\w+)")
```

Comment:
A more precise match for the env var name would be `[A-Za-z_]\w*` as env var names cannot begin with a digit.

And the value should be allowed to be any string (including the empty string) not just `\w+`, which excludes a great many characters. So it should probably be `.*`.

Reply:
Also, would be great to see a unit test for this file, that exercises the regexes, including cases like invalid env var names, and empty values.